### PR TITLE
Chat - Green line is not present correctly on the first unread system message

### PR DESCRIPTION
**Pull request by ikevin127**
### Details

Chat - Green line is not present correctly on the first unread system message. This PR makes sure we always show the new unread message marker above the first unread report action.

In addition we make sure to maintain the manual **Mark as unread** functionality added by PR https://github.com/Expensify/App/pull/29314 which sets the new unread message marker accordingly.
For this I added the testing steps below (6-11) and also demonstrated in the below videos on all platforms.

### Fixed Issues

$ https://github.com/Expensify/App/issues/36399
PROPOSAL: https://github.com/Expensify/App/issues/36399#issuecomment-1941652858

### Tests

**Current issue steps:**
1. User A request money from User B
2. User B open chat with user A
3. User A edit the amount, description, merchant and date
4. User B open the detail IOU page
5. Verify that the new unread message marker appears above the "changed the amount..." report action message.
**Additional steps covering [#29314](https://github.com/Expensify/App/pull/29314) case:**
6. Add a new comment to a chat
7. Mark it as unread
8. Verify that unread mark appears to the new comment
9. Add another comment
10. Mark it as unread
11. Verify that unread mark appears to the other comment

- [x] Verify that no errors appear in the JS console

### Offline tests

**TLDR**: Same as Tests.

**Current issue steps:**
1. User A request money from User B
2. User B open chat with user A
3. User A edit the amount, description, merchant and date
4. User B open the detail IOU page
5. Verify that the new unread message marker appears above the "changed the amount..." report action message.
**Additional steps covering [#29314](https://github.com/Expensify/App/pull/29314) case:**
6. Add a new comment to a chat
7. Mark it as unread
8. Verify that unread mark appears to the new comment
9. Add another comment
10. Mark it as unread
11. Verify that unread mark appears to the other comment

### QA Steps

**TLDR**: Same as Tests.

**Current issue steps:**
1. User A request money from User B
2. User B open chat with user A
3. User A edit the amount, description, merchant and date
4. User B open the detail IOU page
5. Verify that the new unread message marker appears above the "changed the amount..." report action message.
**Additional steps covering [#29314](https://github.com/Expensify/App/pull/29314) case:**
6. Add a new comment to a chat
7. Mark it as unread
8. Verify that unread mark appears to the new comment
9. Add another comment
10. Mark it as unread
11. Verify that unread mark appears to the other comment

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/56457735/14aa8aa1-b258-4639-b75b-9a2574667a82

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/56457735/ea2f2222-5990-4383-9f2d-a54aa5ba0d40

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/56457735/485d2ebf-b33e-4790-b052-30752573e2fb

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/56457735/048c07b4-2fdc-4e11-b952-f33fa0820a2a

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/56457735/24e25a37-d942-482e-b153-68f080972b59

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/56457735/a63b47af-b02e-43fc-8176-7e5691254184

</details>

### CODE REVIEW

**Comment by eVoloshchak** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Comment by eVoloshchak** [comment]
@ikevin127, I tested this yesterday and it wasn't working properly _sometimes_, but there seemed to be some problems with the BE, which I thought were the cause.  But it seems like this does work incorrectly for some cases, I've been able to reliably reproduce one of them

1. User A request money from User B
2. User B open chat with user A
3. User A edit the amount, description
4. User B open the detail IOU page
5. Verify that the new unread message marker appears above the "changed the amount..." report action message 
6. User A, add a new comment to a chat, edit the description
7. User B open the detail IOU page
8. Notice unread message marker is at the top

https://github.com/Expensify/App/assets/9059945/b31f468f-4bbf-469c-aed8-eab4545d3db8

**Comment by ikevin127** [comment]
@eVoloshchak Looking into it, will let you know the outcome soon!

**Comment by ikevin127** [comment]
@eVoloshchak I did some digging and I realized that my initial RCA / solution was wrong.
I updated the proposal, merged w/ main and changed the solution to the one from the [updated proposal](https://github.com/Expensify/App/issues/36399#issuecomment-1941652858).

Turns out the root cause is different, good thing we caught it here and nothing really changes from the PR description in terms of testing steps. I tested extensively and all scenarios pass now!

**Review by eVoloshchak — CHANGES_REQUESTED** [review]
@ikevin127, this is looking good!
Tests good using [these steps](https://github.com/Expensify/App/pull/37082#issuecomment-1969252347), but there's still a small bug:

1. User A request money from User B
2. User B open chat with user A
3. User B open the money request, navigate back (this step is the key to reproducing this bug, it isn't present is you don't visit the newly created money report)
4. User A edit the description and the amount
5. User B open the detail IOU page
6. Green indicator is above "changed amount" message, where it should be above "changed description"

https://github.com/Expensify/App/assets/9059945/13f4eae3-cfdc-4955-8079-d231ca1e5c23

**Comment by ikevin127** [comment]
@eVoloshchak Will look into it today or tomorrow and get back to you, thanks!

**Comment by ikevin127** [comment]
@eVoloshchak I looked into it and I was able to reproduce your scenario about 3 times in 40+ attempts (really hard to reproduce since it's based on how fast you perform the actions one right after the other, at a specific interval).

What happens behind the scene when the line appears above the amount instead of description (this only happens when description and amount are changed really fast one right after the other at a specific interval), we correctly compare `report.lastReadTime` with the action's created date, the logic is working correctly -> we're shown the new unread marker above the amount action because the description's created date is before (earlier) than `report.lastReadTime` -> therefore isMessageUnread returns false based on `report.lastReadTime < action.created` check.

Here’s a screenshot where the amount action change is performed after the description just like in your steps, at the very bottom that date is the lastReadTime of the report, as you can see the description date is before the lastReadTime which in that specific edge case makes the amount action only be new and unread.

Screenshot 2024-03-05 at 00 58 05

Not exactly sure how or why this happens (the description action created date being before the report.lastReadTime), but I merged w/ main and was not able to reproduce your scenario anymore, tried a bunch and it seems to work 100% for me right now.

Please give it a try and hopefully we can move forward with the merge! 

Otherwise I'm out of ideas as to how to tackle this edge case scenario as to me it seems BE (pusher event) related, and I don't think we should hold this issue for that specific edge case if it's still going to be reproducible after this merge w/ main.

**Review by eVoloshchak — COMMENTED** [review]
@ikevin127, testing this right now and seems like something has changed, I'm getting invalid results 100% times, no matter if I edit the description or any other fields
I'm testing this by cherry-picking this PR to the latest main, but there aren't any changes compared to this branch


https://github.com/Expensify/App/assets/9059945/59c090ce-e5b2-4635-8436-fe0dcd621403

Could you merge the latest main and double check if the issue is present for you?

**Comment by ikevin127** [comment]
@eVoloshchak Will look into it and get back to you, thanks!

**Comment by ikevin127** [comment]
@eVoloshchak After thorough testing I came to the conclusion that none of my proposed solutions work in actually fixing the issue, also I was not able to find any other solution to fix the issue.

I think the PR can be closed at this point and the issue should get back to looking for other proposals. The context here can be used by other Contributors which would like to take on the issue.
